src/main.rs

```diff
@@ -99,86 +102,160 @@ struct Args {
+            println!("No preferred mode for the selected connector, exiting");
+            return None;
+        };
 
-    let encoder =
-        gpu.get_encoder(encoder_handle).expect("Failed to get encoder from encoder handle");
-    dbg!(encoder);
+        println!("Using mode: {mode:?}");
 
-    let crtc_handle = *resources
-        .filter_crtcs(encoder.possible_crtcs())
-        .first()
-        .expect("No CRTCs found for encoder");
-    let crtc = gpu.get_crtc(crtc_handle).expect("Failed to get CRTC from CRTC handle");
-    dbg!(crtc);
+        let Some(encoder_handle) = first_encoder(&connector) else {
+            println!("Selected connector does not have an encoder, exiting");
+            return None;
+        };
 
-    let (width, height) = preferred_mode.size();
-    let (width, height) = (width as u32, height as u32);
+        let encoder =
+            gpu.get_encoder(encoder_handle).expect("Failed to get encoder from encoder handle");
+        dbg!(encoder);
+
+        let crtc_handle = *resources
+            .filter_crtcs(encoder.possible_crtcs())
+            .first()
+            .expect("No CRTCs found for encoder");
+        let crtc = gpu.get_crtc(crtc_handle).expect("Failed to get CRTC from CRTC handle");
+        dbg!(crtc);
+
+        let (width, height) = mode.size();
+        let (width, height) = (width as u32, height as u32);
+        let gbm_device = GbmDevice::new(gpu).expect("Failed to create GbmDevice");
+        Some(DrmDisplay { gbm_device, crtc, connector, mode, width, height })
+    }
+}
 
-    let gbm = GbmDevice::new(gpu).expect("Failed to create GbmDevice");
+struct Window {
+    gbm_surface: Surface<BufferObject<()>>,
+    drm_display: DrmDisplay,
+}
 
-    let mut buffer_object = gbm
-        .create_buffer_object::<()>(
-            width,
-            height,
-            BufferFormat::Argb8888,
-            BufferObjectFlags::SCANOUT | BufferObjectFlags::WRITE,
-        )
-        .unwrap();
+impl Window {
+    fn new(drm_display: DrmDisplay) -> Window {
+        let format = BufferFormat::Argb8888;
+        // NOTE(mbernat): This should be used with create_surface() on drivers that support it (not nvidia)
+        // let usage = BufferObjectFlags::SCANOUT | BufferObjectFlags::RENDERING;
+        let modifiers = std::iter::once(Modifier::Linear);
+        let gbm_surface: Surface<BufferObject<()>> = drm_display
+            .gbm_device
+            .create_surface_with_modifiers(drm_display.width, drm_display.height, format, modifiers)
+            .unwrap();
+
+        Window { gbm_surface, drm_display }
+    }
 
-    let buffer_data = vec![255u8; (width * height * 4) as usize];
+    fn swap_buffers(&self) {
+        // TODO(mbernat): move this elsewhere
+        let depth_bits = 24;
+        let bits_per_pixel = 32;
+
+        // SAFETY: this must be called exactly once after `eglSwapBuffers`, which happens in `frame.finish()`.
```

Comment:
Hmm... If this function has expectations on the caller to be used safely, the function itself should be marked `unsafe`. Granted, this is a private function in non-library code, but I think it's still a good habit not to have functions with unenforced invariants.

There might be an alternative way to structure things such that this condition is enforced, but I'm not sure how that would look yet.

Reply:
You're absolutely correct, I'll make it unsafe.

Enforcing this would look similar to glium's `Frame` abstraction: it would swap buffers when you `finish()` it or `drop()` it. In fact, that's exactly what glium does too but only for GL buffers, not GBM buffers :)